### DISCUSSION TITLE
Is there a way for a plugin to run a closure and capture the STDOUT of *Nu* commands?

### DISCUSSION THREAD

**Discussion by idanarye** [nushell/nushell#13079] [Q&A]
[`eval_closure_with_stream`](https://docs.rs/nu-plugin/latest/nu_plugin/struct.EngineInterface.html#method.eval_closure_with_stream) has a `redirect_stdout` argument for redirecting the output of a command - but that only works for external commands. Nu commands that write to STDOUT (mainly using `print`) will still write to STDOUT without having their output captured.

Is there a way around it? I thought about maybe [`Stack::start_capture`](https://docs.rs/nu-protocol/0.94.2/nu_protocol/engine/struct.Stack.html#method.start_capture) would do it, but I think it has to be called by the active stack of the interpreter which I don't now how to acquire (just using `Stack::new().start_capture()` does nothing)

**Comment by ghost** [comment]
AFAIK print is not intended to ever be used by anything afterwards (except tmux/..).  
if you want to pipe data use `echo` (or nothing).

if you want to capture data from a 3rd party script or something i'd either edit it or write a wrapper instead of a plugin.  
if you work with your own code just use echo (`ls | each {|i| [$i.name $i.size]}` instead of `ls | each {|i| print $i.name; print $.size]}` for example).

i haven't worked a lot with plugins lately, but last time i did plugins roughly worked like this:

```
plugin registration
| nu -> plugin: send signature
| plugin -> nu: signature (available commands)
| nu -> file: signature

nu starts
| file -> nu: all signatures
nu executes (and your print happens)
plugin call
| nu crafts request
| nu -> plugin: command + arguments + stdin capture
| plugin executes
| plugin -> nu: result
nu continues
```

unless there was a big change i missed the capture would probably only be of what happens within the plugin (and start capture would have to be called before the plugin is called)

**Comment by idanarye** [comment]
This is my usecase: https://github.com/idanarye/nu_plugin_skim

I'm trying to run [skim](https://github.com/lotabout/skim) as a Nu plugin, so that it could filter Nu types instead of just string. This also means that the `--preview` flag, for example, would accept Nu closures. And if these closures use `print`, I'd want it to get printed to the preview panel instead of directly to the TTY (where it'll interfere with skim's TUI)
